src/ca65/instr.c

```diff
@@ -1538,3 +1707,270 @@ void HandleInstruction (unsigned Index)
+              Emit1(0x2B, A.Expr); /* ADC (zp),y -> JSB2 .db expr */
+              break;
+            case 0xE0: 
+              Emit1(0x3B, A.Expr); /* SBC (zp),y -> JSB3 .db expr */
+              break;
+            case 0x20: 
+              Emit1(0x4B, A.Expr); /* AND (zp),y -> JSB4 .db expr */
+              break;
+            case 0x00: 
+              Emit1(0x5B, A.Expr); /* ORA (zp),y -> JSB5 .db expr */
+              break;
+            case 0x40: 
+              Emit1(0x6B, A.Expr); /* EOR (zp),y -> JSB6 .db expr */
+              break;
+            case 0xC0: 
+              Emit1(0x7B, A.Expr); /* CMP (zp),y -> JSB7 .db expr */
+              break;
+            default:
+              ErrorSkip ("ARITH: Invalid arith base code 0x%x", Ins->BaseCode);
+              return;
+            }
+        }
+      else if (A.AddrModeBit & AM65_DIR_IND)
+        {
+          /* (zp,x) changed to (zp), code using this must be manually fixed with a subroutine.. */
+          Emit1(Ins->BaseCode + 0x01, A.Expr); /* LDA (zp) etc */
+        }
+      else
+        {
+          EmitCode(&A);
+        }         
+    }
+  else
+    {
+      ErrorSkip ("ARITH: invalid adressing mode 0x%lx", A.AddrModeBit);
+      return;
+    }
+}
+
+static void C39PutAdd (const InsDesc* Ins)
+{
+  EffAddr A;
+
+  /* Evaluate the addressing mode used */
+  if (EvalEA (Ins, &A))
+    {
+      if (A.AddrModeBit & AM65_IMM_IMPLICIT)
+        {
+          Emit0(0x89);
+        }
+      else if (A.AddrModeBit & AM65_DIR)
+        {
+          Emit0(0x64);
+        }
+      else if (A.AddrModeBit & AM65_DIR_X)
+        {
+          Emit0(0x74);
+        }
+      else
+        {
+          ErrorSkip ("ADD: Invalid adressing mode 0x%lx", A.AddrModeBit);
+          return;
+        }
+    }
+    
+  EmitByte(A.Expr);
+}
+
+static void C39PutExchange (const InsDesc* Ins)
+{
+  EffAddr A;
+
+  /* Evaluate the addressing mode used */
+  if (EvalEA (Ins, &A))
+    {
+      if (A.AddrModeBit & AM65_DIR_X)
+        {
+          Emit0(0xD4);
+        }
+      else
+        {
+          ErrorSkip ("EXC: Invalid adressing mode 0x%lx", A.AddrModeBit);
+          return;
+        }
+    }
+    
+  EmitByte(A.Expr);
+}
+
+static void C39PutStoreImm (const InsDesc* Ins)
+{
+  ExprNode* immed;
+  ExprNode* addr;
+
+  if (CurTok.Tok != TOK_HASH)
+    {
+      ErrorSkip ("STI: Invalid immed, no # found");
+      return;
+    }
+    
+  NextTok ();
+  
+  immed = Expression ();
+
+  ConsumeComma ();
+
+  addr = Expression();
+
+  Emit0 (Ins->BaseCode);
+  EmitByte (immed);
+  EmitByte (addr);
+}
+
+static void C39PutBARBAS (const InsDesc* Ins)
+{
+    ExprNode* addr;
+    long mask;
+    
+    addr = Expression();
+    ConsumeComma ();
+
+    if (CurTok.Tok != TOK_HASH)
+      {
+        ErrorSkip ("BARBAS: Invalid immed, no # found");
+        return;
+      }
+    
+    NextTok ();
+    
+    /* MUST USE .DEFINE, NOT = FOR CONSTANTS! */
+    mask = ConstExpression();
```

Comment:
How would you handle this case?


On 09.05.2015 05:13, greg-king5 wrote:
>
> In src/ca65/instr.c 
> <https://github.com/cc65/cc65/pull/157#discussion_r29989025>:
>
> > +    ExprNode* addr;
> > +    long mask;
> > +
> > +    addr = Expression();
> > +    ConsumeComma ();
> > +
> > +    if (CurTok.Tok != TOK_HASH)
> > +      {
> > +        ErrorSkip ("BARBAS: Invalid immed, no # found");
> > +        return;
> > +      }
> > +
> > +    NextTok ();
> > +
> > +    /* MUST USE .DEFINE, NOT = FOR CONSTANTS! */
> > +    mask = ConstExpression();
>
> I definitely don't like that! We must be able to use names that were 
> assigned by |=|. But even more important, we must be able to use names 
> that were defined by the |.enum| directive.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cc65/cc65/pull/157/files#r29989025>.
>

-- 
Mit freundlichen Grüssen

Rainer Keuchel